src/nanoemoji/util.py

```diff
@@ -85,30 +87,66 @@ def file_printer(filename):
-    # So for this to work on any TTFont, we first compile everything to a temporary
-    # stream then reload with lazy=False. Input font is modified in-place.
+def require_fully_loaded(font: ttLib.TTFont):
+    not_loaded = sorted(t for t in font.keys() if not font.isLoaded(t))
+    if not_loaded:
+        raise ValueError(f"Everything should be loaded, following aren't: {not_loaded}")
+
+
+def _reload(font: ttLib.TTFont, lazy: bool = True):
+    # Stream font to memory and load it back again
     tmp = BytesIO()
-    ttfont.save(tmp)
+    font.save(tmp)
     tmp.seek(0)
-    ttfont2 = ttLib.TTFont(tmp, lazy=False)
-    for tag in ttfont2.keys():
-        table = ttfont2[tag]
-        # cmap is exceptional in that it always loads subtables lazily upon getting
-        # their attributes, no matter the value of TTFont.lazy option.
-        # TODO: remove this hack once fixed in fonttools upstream
-        if tag == "cmap":
-            _ = [st.cmap for st in table.tables]
-        ttfont[tag] = table
-
-
-def reorder_glyphs(ttfont: ttLib.TTFont, glyph_order: List[str]):
-    # Changing the order of glyphs in a TTFont requires that all tables that use
-    # glyph indexes have been fully decompiled (loaded with lazy=False).
-    # Cf. https://github.com/fonttools/fonttools/issues/2060
-    ensure_ttfont_fully_decompiled(ttfont)
-    ttfont.setGlyphOrder(glyph_order)
-    # glyf table is special and needs its own glyphOrder...
-    assert ttfont.isLoaded("glyf")
-    ttfont["glyf"].glyphOrder = glyph_order
+    return ttLib.TTFont(tmp, lazy=lazy)
+
+
+def load_fully(font: Union[Path, ttLib.TTFont]) -> ttLib.TTFont:
+    if isinstance(font, Path):
+        font = ttLib.TTFont(str(font), lazy=False)
+    else:
+        # A TTFont might be opened lazily and some tables only partially decompiled.
+        # If so, reload it
+        if font.lazy:
+            font = _reload(font, lazy=False)
+
+    font.ensureDecompiled()  # Do what you thought lazy=False meant
+
+    require_fully_loaded(font)
+
+    return font
+
+
+SubTablePath = Tuple[otBase.BaseTable.SubTableEntry, ...]
+
+# Given f(current frontier, new entries) -> new frontier
+AddToFrontierFn = Callable[[List[SubTablePath], List[SubTablePath]], List[SubTablePath]]
+
+
+def dfs_base_table(root: otBase.BaseTable) -> Iterable[SubTablePath]:
+    yield from _traverse_ot_data(root, lambda current, new: new + current)
+
+
+def bfs_base_table(root: otBase.BaseTable) -> Iterable[SubTablePath]:
+    yield from _traverse_ot_data(root, lambda current, new: current + new)
+
+
+def _traverse_ot_data(
+    root: otBase.BaseTable, add_to_frontier_fn: AddToFrontierFn
+) -> Iterable[SubTablePath]:
+    # no visited because general otData is forward-offset only and thus cannot cycle
+
+    frontier: List[SubTablePath] = [(otBase.BaseTable.SubTableEntry("", root),)]
+    while frontier:
+        # path is (value, attr_name) tuples. attr_name is attr of parent to get value
+        path = frontier.pop(0)
```

Comment:
How about using `deque.popleft` which is O(1) instead of list.pop(0) which is O(n) in CPython (lists are array-based) https://docs.python.org/3/library/collections.html#deque-objects

Reply:
Will do